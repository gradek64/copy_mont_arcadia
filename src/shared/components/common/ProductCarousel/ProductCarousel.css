/** @define ProductCarousel */

.ProductCarousel {
  position: relative;
  box-sizing: border-box;
  width: 100%;
  max-width: 800px;
  margin: 0 auto;
  padding: 0 25px;

  &-container {
    overflow: hidden;
    white-space: nowrap;
  }

  &-arrow--left,
  &-arrow--right {
    width: calc($carousel-arrow-width + $carousel-arrow-position);
    height: 100%;
    position: absolute;
    top: 0;
    background: url('public/$brand/images/carousel-arrow-left.svg') 50% no-repeat;
  }

  &-arrow--left {
    left: 0;
  }

  &-arrow--right {
    right: 0;
    transform: scaleX(-1);
  }

  &-arrow--top,
  &-arrow--bottom {
    width: 100%;
    height: $carousel-vertical-arrow-height;
    background: url('public/common/images/carousel-vertical-arrow.png') center no-repeat;
    background-size: $carousel-vertical-arrow-width auto;
    z-index: 1;
    margin: 5px 0;
    flex-shrink: 0;
  }
  &-arrow--bottom {
    transform: rotate(180deg);
  }

  &-vertical {
    width: 150px;
    display: flex;
    flex-direction: column;
    font-size: 15px;
    .ProductCarousel-container {
      white-space: pre-wrap;
      height: 350px;
    }
  }
}

@media $min-laptop {
  .ProductCarousel {
    &-arrow--left,
    &-arrow--right {
      height: 80%;
    }
  }
}
