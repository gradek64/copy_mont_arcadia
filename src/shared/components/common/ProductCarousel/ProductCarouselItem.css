/** @define ProductCarouselItem; weak */

.ProductCarouselItem {
  box-sizing: border-box;
  display: inline-block;
  vertical-align: top;
  position: relative;
  transition-property: opacity, transform;
  transition-duration: 0.3s;
  transition-timing-function: ease;
  background-color: $white;
  white-space: normal;

  & .Product-link {
    padding-top: 0;
  }

  &-delete {
    position: absolute;
    top: 0;
    right: $plp-product-horizontal-padding;
    width: 44px;
    height: 44px;
    color: $carousel-product-remove-color;
    background-color: transparent;
    border-color: transparent;

    &:before {
      content: 'x'
    }
  }

  &-vertical {
    width: 100%;
    display: flex;

    & .Product {
      padding: 10px 0;
      margin: 0;
    }
  }
}
