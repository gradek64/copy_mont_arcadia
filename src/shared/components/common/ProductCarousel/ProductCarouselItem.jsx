import PropTypes from 'prop-types'
import React from 'react'
import { connect } from 'react-redux'

import { getProductRouteFromParams } from '../../../lib/get-product-route'
import { clearProduct } from '../../../actions/common/productsActions'

import { IMAGE_SIZES } from '../../../constants/amplience'
import {
  CAROUSEL_AXIS_HORIZONTAL,
  CAROUSEL_AXIS_VERTICAL,
} from '../../../constants/productCarouselConstants'
import Product from '../Product/Product'

const ProductCarouselItem = (
  {
    lazyLoad,
    productId,
    name,
    price,
    unitPrice,
    salePrice,
    url,
    imageUrl,
    amplienceUrl,
    imageSize,
    style,
    canRemoveProduct,
    storeCode,
    lang,
    clearProduct,
    onProductRemove,
    onLinkClick,
    onQuickViewButtonClick,
    hideProductMeta,
    hideProductName,
    hideQuickViewIcon,
    hidePrice,
    isImageFallbackEnabled,
    openQuickViewOnProductClick,
    axis,
  },
  { l }
) => {
  const linkClickHandler = () => {
    clearProduct()
    onLinkClick(productId)
  }

  return (
    <div
      className={`ProductCarouselItem ProductCarouselItem-${axis}`}
      style={style}
    >
      <Product
        isImageFallbackEnabled={isImageFallbackEnabled}
        lazyLoad={lazyLoad}
        grid={1}
        sizes={imageSize}
        name={name}
        productId={productId}
        productUrl={
          url ||
          (productId
            ? getProductRouteFromParams(lang, storeCode, l`product`, productId)
            : '')
        }
        assets={imageUrl ? [{ url: imageUrl }] : []}
        productBaseImageUrl={amplienceUrl}
        unitPrice={salePrice || unitPrice || price}
        wasPrice={unitPrice}
        showPrice={(!!unitPrice || !!salePrice || !!price) && !hidePrice}
        onLinkClick={linkClickHandler}
        onQuickViewButtonClick={onQuickViewButtonClick}
        showProductView={false}
        openQuickViewOnProductClick={openQuickViewOnProductClick}
        hideProductMeta={hideProductMeta}
        hideProductName={hideProductName}
        hideQuickViewIcon={hideQuickViewIcon}
        isCarouselItem
      />
      {canRemoveProduct && (
        <button
          className="ProductCarouselItem-delete Button Button--secondary"
          onClick={onProductRemove.bind(null, productId)} // eslint-disable-line react/jsx-no-bind
        >
          <span className="screen-reader-text">
            Remove product from carousel
          </span>
        </button>
      )}
    </div>
  )
}

ProductCarouselItem.propTypes = {
  isImageFallbackEnabled: PropTypes.bool,
  lazyLoad: PropTypes.bool,
  productId: PropTypes.number,
  name: PropTypes.string,
  price: PropTypes.string,
  unitPrice: PropTypes.string,
  salePrice: PropTypes.string,
  url: PropTypes.string,
  imageUrl: PropTypes.string,
  amplienceUrl: PropTypes.string,
  imageSize: PropTypes.oneOf(Object.values(IMAGE_SIZES)),
  style: PropTypes.objectOf(PropTypes.string),
  axis: PropTypes.oneOf([CAROUSEL_AXIS_HORIZONTAL, CAROUSEL_AXIS_VERTICAL]),
  canRemoveProduct: PropTypes.bool,
  hideProductMeta: PropTypes.bool,
  hideProductName: PropTypes.bool,
  hideQuickViewIcon: PropTypes.bool,
  hidePrice: PropTypes.bool,
  storeCode: PropTypes.string,
  lang: PropTypes.string,
  onLinkClick: PropTypes.func,
  onQuickViewButtonClick: PropTypes.func,
  onProductRemove: PropTypes.func,
  // actions
  clearProduct: PropTypes.func.isRequired,
}

ProductCarouselItem.defaultProps = {
  isImageFallbackEnabled: false,
  lazyLoad: false,
  productId: undefined,
  name: undefined,
  price: undefined,
  unitPrice: undefined,
  salePrice: undefined,
  url: undefined,
  imageUrl: undefined,
  amplienceUrl: '',
  imageSize: IMAGE_SIZES.smallProduct,
  style: {},
  canRemoveProduct: false,
  hideProductMeta: false,
  hideProductName: false,
  hideQuickViewIcon: false,
  hidePrice: false,
  openQuickViewOnProductClick: false,
  storeCode: undefined,
  lang: undefined,
  axis: CAROUSEL_AXIS_HORIZONTAL,
  onLinkClick: () => {},
  onQuickViewButtonClick: () => {},
  onProductRemove: () => {},
}

ProductCarouselItem.contextTypes = {
  l: PropTypes.func,
}

export default connect(
  (state) => ({
    storeCode: state.config.storeCode,
    lang: state.config.lang,
  }),
  { clearProduct }
)(ProductCarouselItem)

export { ProductCarouselItem as WrappedProductCarouselItem }
