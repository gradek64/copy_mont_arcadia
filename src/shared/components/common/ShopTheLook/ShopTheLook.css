/** @define ShopTheLook */
@lost flexbox flex;

.ShopTheLook {
  &-link {
    line-height: 30px;
    min-height: 0;
    padding-left: 15px;
    padding-right: 15px;
    width: auto;
    display: inline-block;
    text-decoration: none;
  }
}
