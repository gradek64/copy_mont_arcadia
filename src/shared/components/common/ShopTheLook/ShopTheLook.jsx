import PropTypes from 'prop-types'
import React from 'react'
import { Link } from 'react-router'

import { getRouteFromUrl } from '../../../lib/get-product-route'

// Components
import ProductCarousel from '../ProductCarousel/ProductCarousel'

const ShopTheLook = ({ shopTheLookProducts, bundleURL }, { l }) => {
  const carouselProducts = Array.isArray(shopTheLookProducts)
    ? shopTheLookProducts.map(
        ({
          bundleImagePath,
          productDisplayURL,
          bundleProductPrice,
          productId,
        }) => ({
          name: '',
          url: productDisplayURL,
          imageUrl: bundleImagePath,
          price: bundleProductPrice,
          productId: parseInt(productId, 10),
        })
      )
    : []

  return carouselProducts.length ? (
    <div className="ShopTheLook">
      <h3>{l`Shop the set`}</h3>
      <ProductCarousel
        isImageFallbackEnabled
        products={carouselProducts}
        hidePrice
      />
      <Link className="ShopTheLook-link Button" to={getRouteFromUrl(bundleURL)}>
        {l`View The Set`}
      </Link>
    </div>
  ) : null
}

ShopTheLook.propTypes = {
  shopTheLookProducts: PropTypes.oneOfType([PropTypes.bool, PropTypes.array]),
  bundleURL: PropTypes.string.isRequired,
}

ShopTheLook.defaultProps = {
  shopTheLookProducts: [],
  bundleURL: '',
}

ShopTheLook.contextTypes = {
  l: PropTypes.func,
}

export default ShopTheLook
