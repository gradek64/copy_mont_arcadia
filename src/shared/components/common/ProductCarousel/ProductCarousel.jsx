import PropTypes from 'prop-types'
import { mathMod } from 'ramda'
import { connect } from 'react-redux'
import React, { Component } from 'react'
import Hammer from 'react-hammerjs'

import { isMobile } from '../../../selectors/viewportSelectors'
import {
  CAROUSEL_AXIS_HORIZONTAL,
  CAROUSEL_AXIS_VERTICAL,
  CAROUSEL_VERTICAL_SIZE_DESKTOP,
  CAROUSEL_HORIZONTAL_SIZE_MOBILE,
  CAROUSEL_HORIZONTAL_SIZE_DESKTOP,
} from '../../../constants/productCarouselConstants'
import { IMAGE_SIZES } from '../../../constants/amplience'

// components
import ProductCarouselItem from './ProductCarouselItem'

const Arrow = ({ direction, onClick }) => {
  return (
    <button
      className={`ProductCarousel-arrow ProductCarousel-arrow--${direction}`}
      onClick={onClick}
      aria-hidden
    />
  )
}

Arrow.propTypes = {
  direction: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired,
}

// helpers
const doubleArray = (array) => array.concat(array)

const getCarouselSize = (isMobile, axis) => {
  if (axis === CAROUSEL_AXIS_VERTICAL) {
    return CAROUSEL_VERTICAL_SIZE_DESKTOP
  } else if (axis === CAROUSEL_AXIS_HORIZONTAL && isMobile) {
    return CAROUSEL_HORIZONTAL_SIZE_MOBILE
  }

  return CAROUSEL_HORIZONTAL_SIZE_DESKTOP
}

class ProductCarousel extends Component {
  static propTypes = {
    isImageFallbackEnabled: PropTypes.bool,
    carouselSize: PropTypes.number.isRequired,
    products: PropTypes.arrayOf(
      PropTypes.shape({
        productId: PropTypes.number,
        name: PropTypes.string,
        price: PropTypes.string,
        unitPrice: PropTypes.string,
        salePrice: PropTypes.string,
        url: PropTypes.string,
        imageUrl: PropTypes.string,
        amplienceUrl: PropTypes.string,
      })
    ),
    imageSize: PropTypes.oneOf(Object.values(IMAGE_SIZES)),
    canRemoveProduct: PropTypes.bool,
    hideProductMeta: PropTypes.bool,
    hideProductName: PropTypes.bool,
    hideQuickViewIcon: PropTypes.bool,
    hidePrice: PropTypes.bool,
    openQuickViewOnProductClick: PropTypes.bool,
    onProductRemove: PropTypes.func,
    onProductLinkClick: PropTypes.func,
    onQuickViewClick: PropTypes.func,
    axis: PropTypes.oneOf([CAROUSEL_AXIS_HORIZONTAL, CAROUSEL_AXIS_VERTICAL]),
    peepNextItem: PropTypes.bool,
    hammerOptions: PropTypes.object,
    onCarouselArrowClick: PropTypes.func,
  }

  static defaultProps = {
    isImageFallbackEnabled: false,
    products: [],
    canRemoveProduct: false,
    onProductRemove: () => {},
    onProductLinkClick: () => {},
    onQuickViewClick: () => {},
    onCarouselArrowClick: () => {},
    hideProductMeta: false,
    hideProductName: false,
    hideQuickViewIcon: false,
    hidePrice: false,
    imageSize: IMAGE_SIZES.smallProduct,
    openQuickViewOnProductClick: false,
    axis: CAROUSEL_AXIS_HORIZONTAL,
    peepNextItem: false,
    hammerOptions: {},
  }

  state = {
    position: 1,
  }

  handleSwipe = ({ direction } = {}) => {
    const { axis } = this.props
    const BACKWARD = 2
    const FORWARD = 4

    if (axis === CAROUSEL_AXIS_HORIZONTAL) {
      if (direction === BACKWARD) {
        this.moveCarouselForward()
      } else if (direction === FORWARD) {
        this.moveCarouselBackward()
      }
    }
  }

  handleOnCarouselArrowClick = () => {
    const { onCarouselArrowClick } = this.props

    if (typeof onCarouselArrowClick === 'function') {
      onCarouselArrowClick()
    }
  }

  moveCarouselForward = () => {
    this.handleOnCarouselArrowClick()
    this.moveCarousel('forward')
  }

  moveCarouselBackward = () => {
    this.handleOnCarouselArrowClick()
    this.moveCarousel('backward')
  }

  moveCarousel(direction) {
    const { position } = this.state
    const newPostion = direction === 'forward' ? position - 1 : position + 1
    this.setState({
      position: mathMod(newPostion, this.getCarouselItems().length),
    })
  }

  getCarouselItems() {
    const { products, carouselSize } = this.props
    // if there is exactly one more item than the carousel size, 'double' the items (to avoid the wrapping effect)
    return products.length === carouselSize + 1
      ? doubleArray(products)
      : products
  }

  getCarouselArrowsPositions() {
    const { axis } = this.props
    const positions = {
      arrowPosition1: 'left',
      arrowPosition2: 'right',
    }

    if (axis === CAROUSEL_AXIS_VERTICAL) {
      positions.arrowPosition1 = 'top'
      positions.arrowPosition2 = 'bottom'
    }

    return positions
  }

  renderProductCarouselItem = (product, index) => {
    const {
      canRemoveProduct,
      onProductRemove,
      onProductLinkClick,
      onQuickViewClick,
      hideProductMeta,
      hideProductName,
      hideQuickViewIcon,
      hidePrice,
      carouselSize,
      imageSize,
      isImageFallbackEnabled,
      axis,
      openQuickViewOnProductClick,
      peepNextItem,
    } = this.props

    const carouselItems = this.getCarouselItems()
    const style = {}
    const peepCarouselSize = peepNextItem ? carouselSize + 0.3 : carouselSize

    if (axis === CAROUSEL_AXIS_HORIZONTAL) {
      style.width = `${100 / peepCarouselSize}%`
    } else {
      style.height = `${100 / carouselSize}%`
    }

    if (carouselItems.length > carouselSize) {
      const carouselPosition = mathMod(
        this.state.position + index,
        carouselItems.length
      )

      const isVisible = carouselPosition > 0 && carouselPosition <= carouselSize
      const translate = (carouselPosition - index - 1) * 100

      if (axis === CAROUSEL_AXIS_HORIZONTAL) {
        style.transform = `translateX(${translate}%)`
      } else {
        style.transform = `translateY(${translate}%)`
      }
      style.zIndex = isVisible ? '1' : '0'
    }

    return (
      <ProductCarouselItem
        isImageFallbackEnabled={isImageFallbackEnabled}
        key={index} // eslint-disable-line react/no-array-index-key
        lazyLoad
        productId={product.productId}
        name={product.name}
        price={product.price}
        unitPrice={product.unitPrice}
        salePrice={product.salePrice}
        url={product.url}
        imageUrl={product.imageUrl}
        amplienceUrl={product.amplienceUrl}
        imageSize={imageSize}
        style={style}
        axis={axis}
        onLinkClick={onProductLinkClick}
        onQuickViewButtonClick={onQuickViewClick}
        canRemoveProduct={canRemoveProduct}
        onProductRemove={onProductRemove}
        hideProductMeta={hideProductMeta}
        hideProductName={hideProductName}
        hideQuickViewIcon={hideQuickViewIcon}
        hidePrice={hidePrice}
        openQuickViewOnProductClick={openQuickViewOnProductClick}
      />
    )
  }

  render() {
    const carouselItems = this.getCarouselItems()
    const { carouselSize, axis, peepNextItem, hammerOptions } = this.props

    const isCarousel = carouselItems.length > carouselSize
    const { arrowPosition1, arrowPosition2 } = this.getCarouselArrowsPositions()
    const shouldShowArrows = isCarousel && !peepNextItem

    return carouselItems.length ? (
      <div className={`ProductCarousel ProductCarousel-${axis}`}>
        {shouldShowArrows && (
          <Arrow
            direction={arrowPosition1}
            onClick={this.moveCarouselBackward}
          />
        )}
        <Hammer onSwipe={this.handleSwipe} options={hammerOptions}>
          <div className="ProductCarousel-container">
            {carouselItems.map(this.renderProductCarouselItem)}
          </div>
        </Hammer>
        {shouldShowArrows && (
          <Arrow
            direction={arrowPosition2}
            onClick={this.moveCarouselForward}
          />
        )}
      </div>
    ) : null
  }
}

export default connect((state, { axis = CAROUSEL_AXIS_HORIZONTAL }) => ({
  carouselSize: getCarouselSize(isMobile(state), axis),
}))(ProductCarousel)

export { ProductCarousel as WrappedProductCarousel, Arrow }
