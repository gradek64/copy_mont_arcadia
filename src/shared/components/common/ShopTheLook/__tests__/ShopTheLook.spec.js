import { Link } from 'react-router'

import testComponentHelper from 'test/unit/helpers/test-component'

import ShopTheLook from '../ShopTheLook'
import ProductCarousel from '../../ProductCarousel/ProductCarousel'

describe('<ShopTheLook />', () => {
  const shopTheLookProducts = [
    {
      bundleId: '22092706',
      productDisplayURL:
        'http://ts.pplive.arcadiagroup.ltd.uk/en/tsuk/product/sleeveless-suede-trench-coat-by-rare-4699824',
      bundleImagePath:
        'http://ts.pplive.arcadiagroup.ltd.uk/wcsstore/TopShop/images/catalog/62S15IBLK_thumb.jpg',
      bundleProductPrice: '55.0',
      productId: '12345',
    },
    {
      bundleId: '22092706',
      productDisplayURL:
        'http://ts.pplive.arcadiagroup.ltd.uk/en/tsuk/product/skeleton-print-mini-dress-4857523',
      bundleImagePath:
        'http://ts.pplive.arcadiagroup.ltd.uk/wcsstore/TopShop/images/catalog/10D22IBLK_thumb.jpg',
      bundleProductPrice: '28.0',
      productId: '123456',
    },
  ]
  const requiredProps = {
    bundleURL:
      'http://ts.pplive.arcadiagroup.ltd.uk/en/tsuk/product/3-fixed-outfit-shop-the-look-4910692?bundle=true',
  }
  const renderComponent = testComponentHelper(ShopTheLook)

  describe('@renders', () => {
    it('should render nothing by default', () => {
      const { wrapper } = renderComponent(requiredProps)
      expect(wrapper.isEmptyRender()).toBe(true)
    })

    it('should render if there are products', () => {
      const { getTree } = renderComponent({
        ...requiredProps,
        shopTheLookProducts,
      })
      expect(getTree()).toMatchSnapshot()
    })

    it('should pass correct `products` prop to <ProductCarousel />', () => {
      const { wrapper } = renderComponent({
        ...requiredProps,
        shopTheLookProducts,
      })
      expect(wrapper.find(ProductCarousel).prop('products')).toEqual([
        {
          name: '',
          url:
            'http://ts.pplive.arcadiagroup.ltd.uk/en/tsuk/product/sleeveless-suede-trench-coat-by-rare-4699824',
          imageUrl:
            'http://ts.pplive.arcadiagroup.ltd.uk/wcsstore/TopShop/images/catalog/62S15IBLK_thumb.jpg',
          price: '55.0',
          productId: 12345,
        },
        {
          name: '',
          url:
            'http://ts.pplive.arcadiagroup.ltd.uk/en/tsuk/product/skeleton-print-mini-dress-4857523',
          imageUrl:
            'http://ts.pplive.arcadiagroup.ltd.uk/wcsstore/TopShop/images/catalog/10D22IBLK_thumb.jpg',
          price: '28.0',
          productId: 123456,
        },
      ])
    })

    it('should pass correct `to` prop to <Link />', () => {
      const { wrapper } = renderComponent({
        ...requiredProps,
        shopTheLookProducts,
      })
      expect(wrapper.find(Link).prop('to')).toBe(
        '/en/tsuk/product/3-fixed-outfit-shop-the-look-4910692?bundle=true'
      )
    })
  })
})
