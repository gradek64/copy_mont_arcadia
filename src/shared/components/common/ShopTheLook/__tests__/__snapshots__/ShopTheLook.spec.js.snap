// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ShopTheLook /> @renders should render if there are products 1`] = `
<div
  className="ShopTheLook"
>
  <h3>
    Shop the set
  </h3>
  <Connect(ProductCarousel)
    hidePrice={true}
    isImageFallbackEnabled={true}
    products={
      Array [
        Object {
          "imageUrl": "http://ts.pplive.arcadiagroup.ltd.uk/wcsstore/TopShop/images/catalog/62S15IBLK_thumb.jpg",
          "name": "",
          "price": "55.0",
          "productId": 12345,
          "url": "http://ts.pplive.arcadiagroup.ltd.uk/en/tsuk/product/sleeveless-suede-trench-coat-by-rare-4699824",
        },
        Object {
          "imageUrl": "http://ts.pplive.arcadiagroup.ltd.uk/wcsstore/TopShop/images/catalog/10D22IBLK_thumb.jpg",
          "name": "",
          "price": "28.0",
          "productId": 123456,
          "url": "http://ts.pplive.arcadiagroup.ltd.uk/en/tsuk/product/skeleton-print-mini-dress-4857523",
        },
      ]
    }
  />
  <Link
    className="ShopTheLook-link Button"
    onlyActiveOnIndex={false}
    style={Object {}}
    to="/en/tsuk/product/3-fixed-outfit-shop-the-look-4910692?bundle=true"
  >
    View The Set
  </Link>
</div>
`;
